File: 26-Visitor/src/go/evaluate_visitor.go
package main

// EvaluateVisitor computes the numeric value of the expression. Binary nodes
// recursively visit their children and combine the results.
type EvaluateVisitor struct {
	result float64
}

func (e *EvaluateVisitor) Result() float64 { return e.result }

func (e *EvaluateVisitor) VisitNumber(number *NumberExpr) {
	e.result = number.Value()
}

func (e *EvaluateVisitor) VisitAddition(addition *Addition) {
	addition.Left().Accept(e)
	left := e.result
	addition.Right().Accept(e)
	right := e.result
	e.result = left + right
}

func (e *EvaluateVisitor) VisitMultiplication(multiplication *Multiplication) {
	multiplication.Left().Accept(e)
	left := e.result
	multiplication.Right().Accept(e)
	right := e.result
	e.result = left * right
}

File: 26-Visitor/src/rust/evaluate_visitor.rs
use crate::visitor::Visitor;
use crate::number_expr::NumberExpr;
use crate::addition::Addition;
use crate::multiplication::Multiplication;

/// Concrete Visitor — computes the numeric value of the expression.
pub struct EvaluateVisitor {
    pub result: f64,
}

impl EvaluateVisitor {
    pub fn new() -> Self {
        EvaluateVisitor { result: 0.0 }
    }
}

impl Visitor for EvaluateVisitor {
    fn visit_number(&mut self, number: &NumberExpr) {
        self.result = number.value();
    }

    fn visit_addition(&mut self, addition: &Addition) {
        addition.left().accept(self);
        let left = self.result;
        addition.right().accept(self);
        let right = self.result;
        self.result = left + right;
    }

    fn visit_multiplication(&mut self, multiplication: &Multiplication) {
        multiplication.left().accept(self);
        let left = self.result;
        multiplication.right().accept(self);
        let right = self.result;
        self.result = left * right;
    }
}
